### PR TITLE
Centralizes error handling of line and column numbers 

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This changes the lex API to return the line and column number it stopped
at. This removes a lot of redundant formatting code and allow users to
control error formatting centrally.

### CODE REVIEW

File: wasm/wat/lexer_test.go [hunk 1]
```diff
@@ -442,101 +442,127 @@ func TestLex(t *testing.T) {
...
-			input:       []byte(")("),
-			expectedErr: "test.wasm:1:1: found ')' before '('",
+			name:         "open paren at end of input",
+			input:        []byte("("),
+			expectedLine: 1,
+			expectedCol:  1,
+			expectedErr:  "found '(' at end of input",
 		},
 		{
-			name:        "unbalanced nesting",
```
Comment (codefromthecrypt) [inline wasm/wat/lexer_test.go:455]: apologies that removing the paren depth tests misaligned the diff.

**Comment by codefromthecrypt** [comment]
marked as draft until I have representative usage example

**Comment by codefromthecrypt** [comment]
one design I tried and dismissed was having the parser return a result of the next parser. While neat, this slowed benchmarks 5x!

Next step will be to show how to use this API to parse a module iteratively.

**Comment by codefromthecrypt** [comment]
https://github.com/tetratelabs/wazero/pull/78 should go before this, and show that this mostly just helps with error handling. we may not actually need to iteratively parse afterall so I may repurpose this solely to centralize error formatting with line/col

File: wasm/wat/lexer_test.go [hunk 2]
```diff
@@ -571,9 +647,9 @@ func BenchmarkLex(b *testing.B) {
 		b.Run(bm.name, func(b *testing.B) {
 			b.ReportAllocs()
 			for i := 0; i < b.N; i++ {
-				err := lex(noopParseToken, "test.wasm", bm.data)
+				line, col, err := lex(noopParseToken, bm.data)
 				if err != nil {
-					panic(err)
+					panic(fmt.Errorf("%d:%d: %w", line, col, err))
```
Comment (codefromthecrypt) [inline wasm/wat/lexer_test.go:652]: a real caller who knows the filename could put that in the format, also!

**Review by mathetake — APPROVED** [review]

File: wasm/wat/lexer_test.go (hunk 2 above)
Comment (mathetake) [reply wasm/wat/lexer_test.go:652]: 💯

**Comment by codefromthecrypt** [comment]
thanks for the patience with the api thrash @mathetake!

**Pull request merged**